src/main/scala/scorex/core/ModifiersCache.scala

```diff
@@ -0,0 +1,132 @@
+package scorex.core
+
+
+import scorex.core.consensus.HistoryReader
+import scorex.core.utils.ScorexLogging
+import scorex.core.validation.RecoverableModifierError
+
+import scala.collection.mutable
+import scala.util.{Failure, Success}
+
+/**
+  * A cache which is storing persistent modifiers not applied to history yet.
+  * @tparam PMOD - type of a persistent node view modifier (or a family of modifiers).
+  */
+trait ModifiersCache[PMOD <: PersistentNodeViewModifier, H <: HistoryReader[PMOD, _]] {
+  require(maxSize >= 1)
+
+  def size: Int = currentCacheSize
+
+  /**
+    * How many elements are to be stored in the cache
+    */
+  def maxSize: Int
+
+  type K = mutable.WrappedArray[Byte]
+  type V = PMOD
+
+  protected var currentCacheSize: Int = 0
+
+  protected val cache = mutable.Map[K, V]()
+
+  protected val rememberedKeys = mutable.HashSet[K]()
+
+  /**
+    * Defines a best (and application-specific) candidate to be applied.
+    * @param history - an interface to history which could be needed to define a candiate
+    * @return - candidate if it is found
+    */
+  def findCandidateKey(history: H): Option[K]
+
+  protected def onPut(key: K): Unit = {}
+  protected def onRemove(key: K, rememberKey: Boolean): Unit = {}
+
+  /**
+    * A cache element replacement strategy method, which defines a key to remove from cache when it is overfull
+    */
+  protected def keyToRemove(): K
+
+
+  def contains(key: K): Boolean = cache.contains(key) || rememberedKeys.contains(key)
+
+  def put(key: K, value: V): Unit = synchronized {
+    onPut(key)
+    cache.put(key, value)
+    currentCacheSize = currentCacheSize + 1
+
+    if (currentCacheSize > maxSize) remove(keyToRemove())
+  }
+
+  /**
+    * Remove an element from the cache.
+    * @param key - modifier's key
+    * @param rememberKey - whether to remember the key as belonging to cache. E.g. invalid modifiers are
+    *                    to be remembered (for not to be requested from the network again).
+    * @return
+    */
+  def remove(key: K, rememberKey: Boolean = false): Option[V] = synchronized {
+    onRemove(key, rememberKey)
+    currentCacheSize = currentCacheSize - 1
+    if (rememberKey) rememberedKeys += key
+    cache.remove(key)
+  }
+
+  def popCandidate(history: H): Option[V] = synchronized {
+    findCandidateKey(history).flatMap(k => remove(k))
+  }
+}
+
+trait LRUCache[PMOD <: PersistentNodeViewModifier, HR <: HistoryReader[PMOD, _]] extends ModifiersCache[PMOD, HR] {
```

Comment:
Why can't we use LRUCache from some library? This should solve some problems, including `//todo: can we do faster search?`

Reply:
@catena2w what library are you talking about?

Reply:
Possibly https://github.com/blemale/scaffeine or https://cb372.github.io/scalacache/ .

Reply:
scalacache + Google Guava might be the best choice, as soon as we already have it in dependencies.
Guava have 2 options: `maximumSize(1000)` and `expireAfterWrite(10, TimeUnit.MINUTES)` that might both be useful